pkg/endpoint/policy.go

```diff
@@ -498,14 +512,91 @@ func (e *Endpoint) Regenerate(regenMetadata *regeneration.ExternalRegenerationMe
 				e.getLogger().Debug("regeneration was cancelled")
 			}
 		}
-
-		done <- buildSuccess
-		close(done)
 	}()
 
 	return done
 }
 
+var reasonRegenRetry = "retrying regeneration"
+
+// startAsyncRegenerationFailureHandler launches a goroutine which waits for a build of
+// the Endpoint to fail. If the Endpoint is deleted, the goroutine is terminated.
+// If a build fails, a controller is launched which tries to regenerate the
+// Endpoint until it suceeds. Once the controller succeeds, it will not be
+// ran again unless another build failure occurs.
+func (e *Endpoint) startAsyncRegenerationFailureHandler() {
```

Comment:
this could be `startRegenerationFailureHandler`. The caller would need choose between executing this func with `go e.startRegenerationFailureHandler` or without `go`